### PR TITLE
Adding data to the return list for use in other functions

### PR DESCRIPTION
**Pull request by JackStat**
It would be convenient if the xgboost objects retained the original data frame so other functions can use the object to do their work.  In R normally the predict function utilizes this object by default so newData does not have to be provided to generate predictions.  Also, I would like to have it available so I can extend the functionality of my R package to generate model metrics for xgboost models automatically

https://github.com/JackStat/ModelMetrics/issues/13

### CODE REVIEW

**Comment by khotilov** [comment]
The fact that a number of modeling algorithms implemented in R are hoarding their training data doesn't mean that it's a generally good thing. Yes, it makes sense for good old `lm`, `glm` and their relatives when not using large datasets. Those models were frequently used to make sense out of existing data (as opposed to a predictive modeling purpose), so it was kind of natural to keep that data in there.

But for general predictive modeling, I tend to think that this approach promotes bad practices (like making it too easy to use perf metrics based on training data, which is essentially what JackStat/ModelMetrics#13 implies), makes users think less about which data is used for certain calculations, and is a nuisance (sometimes a major one) when data size is large. Keep in mind that one of the primary attractions of xgboost is the ability to work with very large datasets. 

For `xgb.cv`, it's a definite *no* from me, since it allows a `Matrix` for `data`, which would lead to memory bloat, and, besides, the result of this function is not even a model.

It's a bit more nuanced with `xgb.train`, as it only accepts `xgb.DMatrix` as `data`, which is a fairly lightweight R object in terms of copying, as it doesn't copy the actual data but only a pointer to it (the latter also means that the actual data would not be persisted with the model). There's somewhat less chance of memory bloat in here, but it still can happen in subtle ways, e.g., when the original data would be updated, but the old data would not be garbage-collected b/c of the existing pointer to it within a model object, and when you create many models each storing a pointer to its own data. Thus, it's not a good thing to do from my point of view. It might be useful, however, to keep some meta-information about the training data, e.g., feature names.

**Comment by JackStat** [comment]
Well shoot. I see your points. Maybe we could send some smaller objects back in the list like you mentioned to make the predict function a bit more predictable :). If you sent back the label or the label values multi-class predictions could be formatted as a matrix instead of a vector that 99% of the time is converted to a matrix or data.frame. Then if you add in the predicted values or an option to add those (like xgb.cv) I can get what I need.

**Comment by khotilov** [comment]
Could you please explain: why does the world need `auc(xgboostModel)`? I definitely see more arguments against it, but maybe I'm missing something important...
